typescript/ecs/capacity-providers-autoscaling/index.ts

```diff
@@ -0,0 +1,216 @@
+
+// Create cluster with FG capacityProviders enabled
+const vpc = new ec2.Vpc(stack, 'Vpc', { maxAzs: 2 });
+const cluster = new ecs.Cluster(stack, 'FargateCPCluster', {
+  vpc,
+  capacityProviders: ['FARGATE', 'FARGATE_SPOT'],
+});
+
+// Create task def with sample image
+const taskDefinition = new ecs.FargateTaskDefinition(stack, 'TaskDef');
+const container = taskDefinition.addContainer('web', {
+  image: ecs.ContainerImage.fromRegistry("amazon/amazon-ecs-sample"),
+  memoryLimitMiB: 256,
+});
+
+container.addPortMappings({
+  containerPort: 80,
+  protocol: ecs.Protocol.TCP
+});
+
+/*
+ * The following service examples all use Application Autoscaling, with a `minCapacity` and
+ * `maxCapacity` used to set `desiredCount` (specified as a range). We examine how to satisfy
+ * different user stories that leverage capacity providers with on top of autoscaling using the
+ * existing CFN API:
+ *
+ * NOTE: The current behavior of the service if it cannot find spot capacity to place tasks on is to
+ * simply fail. For that reason, any specification for placement on spot should not be treated as a
+ * guarantee.
+ *
+ * We will assume that the range is specified as n:m, where n is the minCapacity and m is the
+ * maxCapacity.
+ *
+ * Case 1: n services placed on on-demand instances, with scaling events triggering placement on
+ * spot capacity only
+ *
+ * Case 1bis: Up to p services placed on on-demand instances (where p >= n), with scaling events
+ * triggering placement on spot only
+ *
+ * Case 2: n services placed on spot instances, with scaling events triggering placement on spot
+ * only (i.e. 100% spot)
+ *
+ * Case 3: n services placed on on-demand instances, with scaling events triggering placement on
+ * spot or on-demand (mixed scaling)
+ */
+
+ // Case 1: n services placed on on-demand instances, with scaling events triggering placement on spot
+const service1 = new ecs.FargateService(stack, 'FargateService-Case1-Scale-Into-Spot', {
+  cluster,
+  taskDefinition,
+  capacityProviderStrategies: [
+    {
+      capacityProvider: 'FARGATE_SPOT',
+      weight: 1,
+    },
+    {
+      capacityProvider: 'FARGATE',
+      base: 5, // corresponds min of count range
+      weight: 0,
+    },
+  ],
+});
+
+// Setup AutoScaling policy
+const scaling1 = service1.autoScaleTaskCount({
+  minCapacity: 5,
+  maxCapacity: 10,
+});
+scaling1.scaleOnCpuUtilization('CpuScaling', {
+  targetUtilizationPercent: 50,
+  scaleInCooldown: cdk.Duration.seconds(60),
+  scaleOutCooldown: cdk.Duration.seconds(60)
+});
+
+// Case 2: n services placed on spot instances, with scaling events triggering placement on spot only (i.e. 100% spot)
+const service2 = new ecs.FargateService(stack, 'FargateService-UserStory2a', {
+  cluster,
+  taskDefinition,
+  capacityProviderStrategies: [
+    {
+      capacityProvider: 'FARGATE_SPOT',
+      weight: 1,
+    },
+    {
+      capacityProvider: 'FARGATE',
+      weight: 0,
```

Comment:
this means 0 `FARGATE` because the base defaults to 0? I guess I'm not clear as to why this second object is needed.